### PR TITLE
add assertShape to validate templateElement

### LINKED ISSUE #10195: Value property of templateElement type missing documentation
Link: https://babeljs.io/docs/en/next/babel-types.html#templateelement

Docs reads: value (required)

According to astExplorer it should be: {raw: string, cooked: string}

Babel also doesn't throw if you construct a templateElement without this object, and produces wrong output.

### PR DESCRIPTION
**Pull request by tanhauhau**
Should we update the `templateElement`'s validator and [update the docs as well](https://github.com/babel/website/pull/2061)?

@sebmck need your advice, because i saw you left a `TODO` there.

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #10195 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
If you run `make build` does it updates the generated flow/typescript types?

**Comment by tanhauhau** [comment]
```js
// index.js.flow
declare function templateElement(value: {raw:string, cooked:string}, tail?: boolean): BabelNodeTemplateElement;

// index.d.ts
export function templateElement(value: {raw:string, cooked:string}, tail?: boolean): TemplateElement;
```

File: packages/babel-types/src/definitions/utils.js [hunk 1]
```diff
@@ -161,6 +161,35 @@ export function assertValueType(type: string): Validator {
   return validate;
 }
 
+export function assertShape(shape: {| [string]: Validator |}): Validator {
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/definitions/utils.js]:
```suggestion
export function assertShape(shape: { [string]: Validator }): Validator {
```

indexers don't work with exact objects.

File: packages/babel-types/src/definitions/es2015.js [hunk 2]
```diff
@@ -537,7 +538,10 @@ defineType("TemplateElement", {
   builder: ["value", "tail"],
   fields: {
     value: {
-      // todo: flatten `raw` into main node
+      validate: assertShape({
+        raw: assertValueType("string"),
+        cooked: assertValueType("string"),
```
Comment (jridgewell) [inline packages/babel-types/src/definitions/es2015.js]: `cooked` may also be `null` in a tagged template:

```js
tag`foo\x`
```

Comment (tanhauhau) [reply packages/babel-types/src/definitions/es2015.js]: @jridgewell made `cooked` optional

**Pull request merged**